torch/csrc/jit/script/module.h

```diff
@@ -105,124 +122,133 @@ struct NamedParameter {
+    return *methods.get(name, "method");
   }
 
   std::shared_ptr<Module> get_module(const std::string& name) const {
-    auto loc = find_member(name, NamedMember::Module);
-    return modules.at(loc).module;
+    return modules.get(name, "module").module;
   }
 
-  NamedMember::Kind find_attribute(const std::string& name) {
-    auto it = members.find(name);
-    if(it == members.end())
-      return NamedMember::None;
-    return it->second.kind;
+  const std::vector<NamedModule>& get_modules() const {
+    return modules.values();
   }
-
-  void dump() const {
-    for(auto entry : members) {
-      std::cout << entry.first << ": " << NamedMember::kind_string(entry.second.kind) << "\n";
-    }
+  const  std::vector<NamedParameter>& get_parameters() const {
+    return parameters.values();
   }
 
-private:
-  size_t find_member(const std::string& name, NamedMember::Kind kind) const  {
-    auto it = members.find(name);
-    if(it == members.end()) {
-      std::stringstream ss;
-      ss << "unknown " << NamedMember::kind_string(kind) << " '" << name << "'";
-      throw std::runtime_error(ss.str());
-    }
-    if(it->second.kind != kind) {
-      std::stringstream ss;
-      ss << "Expected attribute '" << name << "' to be a "
-        << NamedMember::kind_string(kind) << " but found "
-        << NamedMember::kind_string(it->second.kind);
-      throw std::runtime_error(ss.str());
-    }
-    JIT_ASSERT(it != members.end() && it->second.kind == kind);
-    return it->second.offset;
+
+  at::optional<NamedParameter&> find_parameter(const std::string& name) {
+    return parameters.find(name);
   }
-  void add_member(const std::string& name, NamedMember::Kind kind, size_t offset) {
-    auto it = members.find(name);
-    if(it != members.end()) {
-      std::stringstream ss;
-      ss << "attempting to add " << NamedMember::kind_string(kind) << " '" << name << "' but Module already contains "
-      << NamedMember::kind_string(it->second.kind) << " '" << name << "'";
-      throw std::runtime_error(ss.str());
-    }
-    members[std::move(name)] = NamedMember { kind, offset };
+  at::optional<NamedModule&> find_module(const std::string& name) {
+    return modules.find(name);
+  }
+  at::optional<Method&> find_method(const std::string& name) {
+    if(auto pm = methods.find(name))
+      return at::optional<Method&>(**pm);
+    return at::nullopt;
   }
+
+private:
+
   // invariant: to ensure member_inputs of Methods stay valid,
   // it is only legal to _add_ new modules and parameters.
   // removing them will allow member_inputs to point to invalid parameters
   // no such restriction exists for methods
-  std::vector<NamedModule> modules;
-  std::vector<NamedParameter> parameters;
-  std::vector<std::unique_ptr<Method>> methods;
-
-  std::unordered_map<std::string, NamedMember> members;
+  OrderedDict<NamedModule> modules;
+  OrderedDict<NamedParameter> parameters;
+  OrderedDict<std::unique_ptr<Method>> methods;
```

Comment:
Can you remind me why do we keep `unique_ptr`s here? Is it because `methods` are quite heavyweight and we want to avoid the cost of moving?

Reply:
Methods are owned by their  module, but we return them as references that are held by python (and tied to the lifetime of the Module). We need to ensure they do not get relocated (e.g. by resizing the vector).